### DISCUSSION TITLE
Face recognition in photos shared by several users, face data not shared?

### DISCUSSION THREAD

**Discussion by Ra72xx** [nextcloud/recognize#1063] [General]
Even if photos are in shared folders and therefore accessible for several users, the face recognition seems to run separately for every single user. Training data, the names of the persons on the photos etc. seem to be only applied on a per user data, i.e. every user has to set up his/her face data individually even if the photos are the same.
This is at least the case for my family photo collection on an local external folder embedded in Nextcloud, which makes the face recognition a bit useless as everybody would have to train the KI with thousands of photos and a lot of false recognitions.
I don't know if this is also the case for the other recgnoze options, I had the impression that object recognition etc. is run only once per shared file and not per user.
I also don't know how to properly handle sharing face recognition data for shared users, but IMHO it should be done somehow, as this sharing of information is the basic idea of Nextcloud.
Or am I understanding something wrong hiere?

**Comment by Sprinterfreak** [comment]
I've come to the same conclusion. Part of the answer is maybe, that if the results are shared, it will be an issue that people can see private results from other users. Or at least can see who knows whom.

But yes. Recognize has to process all data for each user separately. Which is silly.
I've observed this using groupfolders.

**Comment by Olen** [comment]
After a quick verification here https://github.com/nextcloud/recognize/discussions/754#discussioncomment-8092167 I at least got a confirmation that this can be done directly in the database.  This could be a possible solution for a one-time "sync".  Later additions to the library would need to be handled manually by each user (or the query can be run again, of course).

Let's say that _user-a_ has moved all photos of _Person 1_ into a cluster called "Person 1"
_user-b_ has created a similar cluster called "Person One" with some photos in it, but not yet gone through all the shared images that might still be in other clusters, or is still unclustered.

So let's add all the entries that _user-a_ has added to his "Person 1" cluster to _user-b_ s "Person One"-cluster:
```
WITH from_user_cluster AS (
    SELECT id, user_id
    FROM oc_recognize_face_clusters
    WHERE user_id='user-a'
    AND title='Person 1'
), to_user_cluster AS (
    SELECT id, user_id
    FROM oc_recognize_face_clusters
    WHERE user_id='user-b'
    AND title='Person One'
), subquery AS (
    SELECT oc_recognize_face_detections.file_id, oc_recognize_face_detections.x, oc_recognize_face_detections.y, oc_recognize_face_detections.vector
    FROM oc_recognize_face_detections, from_user_cluster
    WHERE oc_recognize_face_detections.user_id=from_user_cluster.user_id
    AND oc_recognize_face_detections.cluster_id=from_user_cluster.id
)
UPDATE oc_recognize_face_detections
SET cluster_id=to_user_cluster.id
FROM from_user_cluster, to_user_cluster, subquery
WHERE oc_recognize_face_detections.user_id=to_user_cluster.user_id
AND oc_recognize_face_detections.file_id=subquery.file_id
AND oc_recognize_face_detections.x=subquery.x
AND oc_recognize_face_detections.y=subquery.y
AND oc_recognize_face_detections.vector=subquery.vector;
```
This query works for *postgresql*, feel free to create similar queries for other database-solutions.

Make sure you replace the `user_id` and `title` in the two first queries. 

I believe the addition of `x`, `y` and `vector` should give a unique match.

Use at your own risk!

**Comment by marcelklehr** [comment]
Also see #948 and #881
